File: src/delta.ts
import { DeltaMagicError } from "./errors.js";

export type DeltaMagic = "kvdiffhp1" | "hdiff13" | "bsdiff40" | "vcdiff" | "unknown";

const HDIFF13 = Buffer.from("HDIFF13&", "ascii");
const KVDIFF = Buffer.from("KVDIFFHP1\n", "ascii");
const BSDIFF = Buffer.from("BSDIFF40", "ascii");
const VCDIFF = Buffer.from([0xd6, 0xc3, 0xc4]);

export function inspectDeltaMagic(bytes: Uint8Array): DeltaMagic {
  if (startsWith(bytes, KVDIFF)) return "kvdiffhp1";
  if (startsWith(bytes, HDIFF13)) return "hdiff13";
  if (startsWith(bytes, BSDIFF)) return "bsdiff40";
  if (startsWith(bytes, VCDIFF)) return "vcdiff";
  return "unknown";
}

export function algoMatchesMagic(algo: string, magic: DeltaMagic): boolean {
  const a = algo.toLowerCase();
  if (a === "hdiffpatch") return magic === "kvdiffhp1" || magic === "hdiff13";
  if (a === "bsdiff") return magic === "bsdiff40";
  if (a === "xdelta3") return magic === "vcdiff";
  return false;
}

export function assertDeltaMagic(algo: string, delta: Uint8Array): DeltaMagic {
  const magic = inspectDeltaMagic(delta);
  if (magic === "unknown") {
    throw new DeltaMagicError("unknown delta container magic");
  }
  if (!algoMatchesMagic(algo, magic)) {
    throw new DeltaMagicError(`delta magic ${magic} does not match algo ${algo}`);
  }
  return magic;
}

function startsWith(bytes: Uint8Array, prefix: Uint8Array): boolean {
  if (bytes.length < prefix.length) return false;
  for (let i = 0; i < prefix.length; i++) {
    if (bytes[i] !== prefix[i]) return false;
  }
  return true;
}

File: src/delta.cpp
#include "delta.hpp"

#include <cstring>

namespace kirivers {

DeltaMagic detect_delta_magic(const Bytes& delta) {
  auto starts = [&](const char* lit) {
    const size_t n = std::strlen(lit);
    if (delta.size() < n) return false;
    return std::memcmp(delta.data(), lit, n) == 0;
  };
  if (starts("KVDIFFHP1\n")) return DeltaMagic::KvDiffHp1;
  if (starts("HDIFF13&")) return DeltaMagic::Hdiff13;
  if (starts("BSDIFF40")) return DeltaMagic::Bsdiff40;
  if (delta.size() >= 3 && static_cast<unsigned char>(delta[0]) == 0xD6 &&
      static_cast<unsigned char>(delta[1]) == 0xC3 &&
      static_cast<unsigned char>(delta[2]) == 0xC4) {
    return DeltaMagic::Vcdiff;
  }
  return DeltaMagic::Unknown;
}

bool magic_matches_algo(DeltaMagic magic, const std::string& algo) {
  if (algo == "hdiffpatch") {
    return magic == DeltaMagic::Hdiff13 || magic == DeltaMagic::KvDiffHp1;
  }
  if (algo == "bsdiff") return magic == DeltaMagic::Bsdiff40;
  if (algo == "xdelta3") return magic == DeltaMagic::Vcdiff;
  return false;
}

}  // namespace kirivers
